### ISSUE TITLE
ListAllWhenCases new rule

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#3176]
## Expected Behavior of the rule
Given
```kotlin
enum class Foo { A, B, C }
```
These should pass:
```kotlin
fun decide(foo: Foo): String =
    when(foo) {
        A -> "A"
        B -> "B"
        C -> "C"
    }

fun decideNullable(foo: Foo?): String =
    when(foo) {
        A -> "A"
        B -> "B"
        C -> "C"
        null -> "null"
    }
```
These should fail:
```kotlin
fun decide(foo: Foo): String =
    when(foo) {
        A -> "A"
        B -> "B"
        else -> "C" // else means C now, but could mean C or D in the future
    }

fun decideNullable(foo: Foo?): String =
    when(foo) {
        null -> "null"
        A -> "A"
        B -> "B"
        else -> "C" // else means C now, but could mean C or D in the future
    }

fun decideNullable2(foo: Foo?): String =
    when(foo) {
        A -> "A"
        B -> "B"
        C -> "C"
        else -> // else means null now, but could mean null or D in the future
    }
```

Configuration parameters:
 * enable for enums
 * enable for sealed classes

Default: off to preserve compatibility, also this my be considered a team style decision.

Suggested group: error prone, which means that maybe it's worth enabling by default?

Note: this is not the same as the related `MissingWhenCase` or `RedundantElseInWhen`, although `MissingWhenCase` could be amended with some configuration parameters to treat `else` as "non-existent" and flag the missing cases. 

## Context
When systems evolve things change, this is true for internal enums too. Imagine you have that `Foo` enum like above and someone adds a `D` constant. In this case if the `when` is listing all cases without `else`, then whoever adds this code will need to fix a compile error and make a decision on what to do exactly with `D`. While if there's an `else`, they'll get some random implicit behavior depending on how the previous developer coded the `when`, which may not be valid for the meaning of `D`.

**Comment by cortinico** [comment]
> although `MissingWhenCase` could be amended with some configuration parameters to treat `else` as "non-existent" and flag the missing cases.

+1 on this as the way to go. Ideally having an `allowElseExpression: true (default)` would be fairly easy to add to control the logic here:

https://github.com/detekt/detekt/blob/7620ce71e3ff055bb65497d56a959b8baa54024f/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCase.kt#L80

**Comment by TWiStErRob** [comment]
Awesome collaboration here👏, thank you @sowmyav24 for doing the heavy lifting.

**Comment by TWiStErRob** [comment]
Hmm the release notes say there's a new `ListAllWhenCases` in 1.15.0, added by https://github.com/detekt/detekt/pull/3297, you might want to remove that as in the end there was new rule added in this issue.

**Comment by cortinico** [comment]
> Hmm the release notes say there's a new `ListAllWhenCases` in 1.15.0, added by #3297, you might want to remove that as in the end there was new rule added in this issue.

Good point https://github.com/detekt/detekt/pull/3856
